clients/azuredevopsrepo/client.go

```diff
@@ -0,0 +1,216 @@
+	if err != nil {
+		return false, fmt.Errorf("could not get repository with error: %w", err)
+	}
+
+	return *repo.IsDisabled, nil
+}
+
+func (client *Client) ListFiles(predicate func(string) (bool, error)) ([]string, error) {
+	return []string{}, nil
+}
+
+func (client *Client) LocalPath() (string, error) {
+	return "", nil
+}
+
+func (client *Client) GetFileReader(filename string) (io.ReadCloser, error) {
+	return nil, nil
+}
+
+func (client *Client) GetBranch(branch string) (*clients.BranchRef, error) {
+	return nil, nil
+}
+
+func (client *Client) GetCreatedAt() (time.Time, error) {
+	return time.Time{}, nil
+}
+
+func (client *Client) GetDefaultBranchName() (string, error) {
+	if len(client.repourl.defaultBranch) > 0 {
+		return client.repourl.defaultBranch, nil
+	}
+
+	return "", fmt.Errorf("%s", "default branch name is empty")
+}
+
+func (client *Client) GetDefaultBranch() (*clients.BranchRef, error) {
+	return client.branches.getDefaultBranch()
+}
+
+func (client *Client) GetOrgRepoClient(context.Context) (clients.RepoClient, error) {
+	return nil, fmt.Errorf("GetOrgRepoClient (GitLab): %w", clients.ErrUnsupportedFeature)
+}
+
+func (client *Client) ListCommits() ([]clients.Commit, error) {
+	return client.commits.listCommits()
+}
+
+func (client *Client) ListIssues() ([]clients.Issue, error) {
+	return nil, nil
+}
+
+func (client *Client) ListLicenses() ([]clients.License, error) {
+	return nil, nil
+}
+
+func (client *Client) ListReleases() ([]clients.Release, error) {
+	return nil, nil
+}
+
+func (client *Client) ListContributors() ([]clients.User, error) {
+	return nil, nil
+}
+
+func (client *Client) ListSuccessfulWorkflowRuns(filename string) ([]clients.WorkflowRun, error) {
+	return nil, nil
+}
+
+func (client *Client) ListCheckRunsForRef(ref string) ([]clients.CheckRun, error) {
+	return nil, nil
+}
+
+func (client *Client) ListStatuses(ref string) ([]clients.Status, error) {
+	return nil, nil
+}
+
+func (client *Client) ListWebhooks() ([]clients.Webhook, error) {
+	return nil, nil
+}
+
+func (client *Client) ListProgrammingLanguages() ([]clients.Language, error) {
+	return nil, nil
+}
+
+func (client *Client) Search(request clients.SearchRequest) (clients.SearchResponse, error) {
+	return clients.SearchResponse{}, nil
+}
+
+func (client *Client) SearchCommits(request clients.SearchCommitsOptions) ([]clients.Commit, error) {
+	return nil, nil
+}
+
+func (client *Client) Close() error {
+	return nil
+}
+
+func CreateAzureDevOpsClient(ctx context.Context, repo clients.Repo) (*Client, error) {
+	token := os.Getenv("AZURE_DEVOPS_AUTH_TOKEN")
+	return CreateAzureDevOpsClientWithToken(ctx, token, repo)
+}
+
+func CreateAzureDevOpsClientWithToken(ctx context.Context, token string, repo clients.Repo) (*Client, error) {
+	// https://dev.azure.com/<org>
+	url := "https://" + repo.Host() + "/" + strings.Split(repo.Path(), "/")[0]
```

Comment:
So the GitHub client doesn't need a repo because it's always `github.com` (unless you set a specific env var), the GitLab client takes a host.

It seems Azure needs an `organizationUrl`. Do you have thoughts on accepting that as a parameter instead of a repo? Or does this give you more flexibility in the future?

Reply:
I do need an `organizationUrl`, but I use `repo` here to add support for on-premises Azure DevOps instances. The host is always `dev.azure.com` for hosted Azure DevOps, but it can be anything for on-premises.